pkg/api/rest/delete.go

```diff
@@ -40,14 +45,17 @@ type RESTDeleteStrategy interface {
 // condition cannot be checked or the gracePeriodSeconds is invalid. The options argument may be updated with
 // default values if graceful is true.
 func BeforeDelete(strategy RESTDeleteStrategy, ctx api.Context, obj runtime.Object, options *api.DeleteOptions) (graceful, gracefulPending bool, err error) {
-	if strategy == nil {
-		return false, false, nil
-	}
-	objectMeta, _, kerr := objectMetaAndKind(strategy, obj)
+	objectMeta, gvk, kerr := objectMetaAndKind(strategy, obj)
 	if kerr != nil {
 		return false, false, kerr
 	}
-
+	if options.Preconditions.UID != nil && *options.Preconditions.UID != objectMeta.UID {
+		return false, false, errors.NewConflict(unversioned.GroupResource{gvk.Group, gvk.Kind}, objectMeta.Name, fmt.Errorf("the UID in the precondition (%s) does not match the UID in record (%s). The object might have been deleted and then recreated", *options.Preconditions.UID, objectMeta.UID))
```

Comment:
This is not sufficient! Another delete/write cycle could happen between this and the actual delete. After checking this, you need to also drive a precondition down into etcd, of either resource version, or entire object body (as read here). If that fails, you then need to do it all over again in a loop (as it may have changed in a way that didn't violate the precondition).